### PR TITLE
simplify EstimateBits a bit

### PR DESCRIPTION
**Pull request by jonsneyers**
Simplify the function EstimateBits a little bit: instead of first rounding up the non-zero counts that would otherwise end up getting a probability below 1/4096, just make all probabilities at least 1/4096 — that's wrong if the count was zero but it doesn't matter since we'll then multiply the number by zero anyway.

This gives a small (~2% or so) speedup at most of the lossless effort settings, but a ~20% speedup at e9.

Both ways of doing the estimate are not quite exact, and this approximation is slightly different from the previous one. The effect of this change on density is small but this way of doing it also seems slightly better for compression (but it's image-dependent and mostly just some fluctuation).

Before (jyrki31):
```
Encoding      kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm  SSIMULACRA2        pnorm       BPP*pnorm   Bugs
-------------------------------------------------------------------------------------------------------------------------
jxl:d0:2        13270 18711689   11.2802146  13.107  17.900         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:3        13270 17219213   10.3804855   6.004   7.038         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:4        13270 17162131   10.3460740   1.635   5.744         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:5        13270 16965129   10.2273127   0.903   4.511         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:6        13270 16901229   10.1887910   0.620   4.155         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:7        13270 16711883   10.0746451   0.440   3.946         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:8        13270 16617202   10.0175673   0.156   3.874         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:9        13270 16447585    9.9153148   0.030   3.836         -nan 100.00000000   0.00000000  0.000000000000      0
```

After:
```
Encoding      kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm  SSIMULACRA2        pnorm       BPP*pnorm   Bugs
-------------------------------------------------------------------------------------------------------------------------
jxl:d0:2        13270 18711689   11.2802146  12.668  18.936         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:3        13270 17219213   10.3804855   5.918   7.102         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:4        13270 17161706   10.3458178   1.646   5.761         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:5        13270 16963847   10.2265399   0.921   4.488         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:6        13270 16898265   10.1870042   0.648   4.136         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:7        13270 16708943   10.0728727   0.463   3.913         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:8        13270 16611499   10.0141293   0.160   3.843         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:9        13270 16447225    9.9150978   0.038   3.807         -nan 100.00000000   0.00000000  0.000000000000      0
```

Before: (nonphoto)
```
Encoding      kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm  SSIMULACRA2        pnorm       BPP*pnorm   Bugs
-------------------------------------------------------------------------------------------------------------------------
jxl:d0:2        14727 12846550    6.9783291  11.368  16.932         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:3        14727 10968310    5.9580570   5.842   7.002         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:4        14727 10868840    5.9040242   1.731   5.523         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:5        14727 10037654    5.4525186   0.874   4.502         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:6        14727  9843434    5.3470171   0.671   4.222         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:7        14727  9458372    5.1378489   0.493   3.978         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:8        14727  9287075    5.0447993   0.200   3.922         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:9        14727  8650805    4.6991733   0.038   3.761         -nan 100.00000000   0.00000000  0.000000000000      0
```

After:
```
Encoding      kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm  SSIMULACRA2        pnorm       BPP*pnorm   Bugs
-------------------------------------------------------------------------------------------------------------------------
jxl:d0:2        14727 12846550    6.9783291  11.236  18.020         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:3        14727 10968310    5.9580570   5.788   7.173         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:4        14727 10871521    5.9054806   1.729   5.513         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:5        14727 10037599    5.4524888   0.888   4.513         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:6        14727  9842695    5.3466156   0.683   4.239         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:7        14727  9456181    5.1366587   0.503   3.983         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:8        14727  9272555    5.0369119   0.205   3.878         -nan 100.00000000   0.00000000  0.000000000000      0
jxl:d0:9        14727  8654170    4.7010012   0.045   3.725         -nan 100.00000000   0.00000000  0.000000000000      0
```

### CODE REVIEW

File: lib/jxl/modular/encoding/enc_ma.cc [hunk 1]
```diff
@@ -37,33 +37,24 @@ const HWY_FULL(float) df;
...
-    const auto round_counts_v = LoadU(di, &rounded_counts[i]);
-    const auto probs = Mul(ConvertTo(df, round_counts_v), inv_total);
-    const auto nbps = IfThenElse(Eq(round_counts_v, total_v), BitCast(di, zero),
-                                 BitCast(di, FastLog2f(df, probs)));
-    bits_lanes = Sub(bits_lanes, IfThenElse(Eq(counts_v, zero), zero,
-                                            Mul(counts_v, BitCast(df, nbps))));
+    const auto counts_iv = LoadU(di, &counts[i]);
+    const auto counts_fv = ConvertTo(df, counts_iv);
+    const auto probs = Mul(counts_fv, inv_total);
+    const auto mprobs = IfThenElse(Lt(probs, minprob), minprob, probs);
```
Comment (veluca93) [inline lib/jxl/modular/encoding/enc_ma.cc:53]: There's Max in hwy for this :)

Comment (jonsneyers) [reply lib/jxl/modular/encoding/enc_ma.cc]: Ah, nice, even simpler then. Updated.

File: lib/jxl/modular/encoding/enc_ma.cc [hunk 2]
```diff
@@ -37,33 +37,24 @@ const HWY_FULL(float) df;
...
-                                 BitCast(di, FastLog2f(df, probs)));
-    bits_lanes = Sub(bits_lanes, IfThenElse(Eq(counts_v, zero), zero,
-                                            Mul(counts_v, BitCast(df, nbps))));
+    const auto counts_iv = LoadU(di, &counts[i]);
+    const auto counts_fv = ConvertTo(df, counts_iv);
+    const auto probs = Mul(counts_fv, inv_total);
+    const auto mprobs = Max(probs, minprob);
+    const auto nbps = IfThenElse(Eq(counts_iv, total_v), BitCast(di, zero),
+                                 BitCast(di, FastLog2f(df, mprobs)));
+    bits_lanes = Sub(bits_lanes, IfThenElse(Eq(counts_fv, zero), zero,
```
Comment (veluca93) [inline lib/jxl/modular/encoding/enc_ma.cc:56]: I think we don't need this "if 0" anymore - log(min_prob) is no longer NaN.

Comment (jonsneyers) [reply lib/jxl/modular/encoding/enc_ma.cc]: Ah right, well spotted, that should make it a little bit faster again then..

**Review by veluca93 — APPROVED** [review]
Can you give updated speed numbers? Thanks!

**Comment by jonsneyers** [comment]
> Can you give updated speed numbers? Thanks!

I couldn't really measure a big difference after removing that unnecessary "if 0", for jyrki31 the numbers are about the same, for the nonphoto corpus I got 0.046 Mpx/s at e9 instead of 0.045 but I didn't measure accurately enough to consider that an actual difference (just one encode repetition, that already takes a while when doing e9 on a set of images).

**Review by veluca93 — DISMISSED** [review]

**Pull request merged**